drivers/hid/hid-nintendo.c

```diff
@@ -1354,6 +1377,44 @@ static void joycon_parse_report(struct joycon_ctlr *ctlr,
 		input_report_key(dev, BTN_EAST, btns & JC_BTN_A);
 		input_report_key(dev, BTN_SOUTH, btns & JC_BTN_B);
 	}
+	if (jc_type_is_nescon(ctlr) || \
+	    jc_type_is_snescon(ctlr) || \
+	    jc_type_is_gencon(ctlr)) {
+ 		s8 x;
+ 		s8 y;
+
+ 		if (btns & JC_BTN_LEFT)
+ 			x = -1;
+ 		else if (btns & JC_BTN_RIGHT)
+ 			x = 1;
+
+ 		if (btns & JC_BTN_UP)
+ 			y = -1;
+ 		else if (btns & JC_BTN_DOWN)
+ 			y = 1;
+
+ 		input_report_abs(dev, ABS_HAT0X, x);
+ 		input_report_abs(dev, ABS_HAT0Y, y);
+
+ 		/* report buttons */
+		input_report_key(dev, BTN_EAST, btns & JC_BTN_A);
+ 		input_report_key(dev, BTN_SOUTH, btns & JC_BTN_B);
+ 		input_report_key(dev, BTN_TL, btns & JC_BTN_L);
+ 		input_report_key(dev, BTN_TR, btns & JC_BTN_R);
+ 		input_report_key(dev, BTN_SELECT, btns & JC_BTN_MINUS);
+ 		input_report_key(dev, BTN_START, btns & JC_BTN_PLUS);
+
+		if (jc_type_is_snescon(ctlr)) {
+ 			input_report_key(dev, BTN_TL2, btns & JC_BTN_ZL);
+ 			input_report_key(dev, BTN_TR2, btns & JC_BTN_ZR);
+ 			input_report_key(dev, BTN_NORTH, btns & JC_BTN_X);
+ 			input_report_key(dev, BTN_WEST, btns & JC_BTN_Y);
+ 		}
+		if (jc_type_is_gencon(ctlr)) {
+ 			input_report_key(dev, BTN_TR2, btns & JC_BTN_ZR);
+ 			input_report_key(dev, BTN_WEST, btns & JC_BTN_Y);
```

Comment:
This was a guess. I need to verify for accuracy. I may make changes in a future commit.